src/platform/i386/chal_pgtbl.c

```diff
@@ -0,0 +1,533 @@
+		/* We need to deact children first! */
+		cos_throw(err, -EINVAL);
+	}
+
+	if (parent == NULL) {
+		if (!root) cos_throw(err, -EINVAL);
+		/**
+                 * Last reference to the captbl page. Require pgtbl
+		 * and cos_frame cap to release the kmem page.
+                 */
+		ret = kmem_deact_pre(deact_header, t, pgtbl_cap, cosframe_addr, &pte, &old_v);
+		if (ret) cos_throw(err, ret);
+	} else {
+		/* more reference exists. */
+		if (root) cos_throw(err, -EINVAL);
+		assert(!pgtbl_cap && !cosframe_addr);
+	}
+
+	if (cos_cas((unsigned long *)&deact_cap->refcnt_flags, l, CAP_MEM_FROZEN_FLAG) != CAS_SUCCESS)
+		cos_throw(err, -ECASFAIL);
+
+	/**
+         * deactivation success. We should either release the
+	 * page, or decrement parent cnt.
+         */
+	if (parent == NULL) {
+		/* move the kmem to COSFRAME */
+		ret = kmem_deact_post(pte, old_v);
+		if (ret) {
+			cos_faa((int *)&deact_cap->refcnt_flags, 1);
+			cos_throw(err, ret);
+		}
+	} else {
+		cos_faa((int *)&parent->refcnt_flags, -1);
+	}
+
+	/* FIXME: this should be before the kmem_deact_post */
+	ret = cap_capdeactivate(dest_ct_cap, capin, CAP_PGTBL, lid);
+	if (ret) cos_throw(err, ret);
+
+	return 0;
+err:
+	return ret;
+}
+
+int
+chal_pgtbl_mapping_add(pgtbl_t pt, u32_t addr, u32_t page, u32_t flags)
+{
+	int                ret = 0;
+	struct ert_intern *pte;
+	u32_t              orig_v, accum = 0;
+
+	assert(pt);
+	assert((PGTBL_FLAG_MASK & page) == 0);
+	assert((PGTBL_FRAME_MASK & flags) == 0);
+
+	/* get the pte */
+	pte = (struct ert_intern *)__pgtbl_lkupan((pgtbl_t)((u32_t)pt | X86_PRESENT), addr >> PGTBL_PAGEIDX_SHIFT,
+	                                          PGTBL_DEPTH, &accum);
+	if (!pte) return -ENOENT;
+	orig_v = (u32_t)(pte->next);
+
+	if (orig_v & X86_PRESENT) return -EEXIST;
+	if (orig_v & X86_COSFRAME) return -EPERM;
+
+	/* Quiescence check */
+	ret = pgtbl_quie_check(orig_v);
+	if (ret) return ret;
+
+	/* ref cnt on the frame. */
+	ret = retypetbl_ref((void *)page);
+	if (ret) return ret;
+
+	ret = __pgtbl_update_leaf(pte, (void *)(page | flags), orig_v);
+	/* restore the refcnt if necessary */
+	if (ret) retypetbl_deref((void *)page);
+
+	return ret;
+}
+
+int
+chal_pgtbl_cosframe_add(pgtbl_t pt, u32_t addr, u32_t page, u32_t flags)
+{
+	struct ert_intern *pte;
+	u32_t              orig_v, accum = 0;
+
+	assert(pt);
+	assert((PGTBL_FLAG_MASK & page) == 0);
+	assert((PGTBL_FRAME_MASK & flags) == 0);
+
+	/* get the pte */
+	pte    = (struct ert_intern *)__pgtbl_lkupan((pgtbl_t)((u32_t)pt | X86_PRESENT), addr >> PGTBL_PAGEIDX_SHIFT,
+                                                  PGTBL_DEPTH, &accum);
+	orig_v = (u32_t)(pte->next);
+	assert(orig_v == 0);
+
+	return __pgtbl_update_leaf(pte, (void *)(page | flags), 0);
+}
+
+/* This function updates flags of an existing mapping. */
+int
+chal_pgtbl_mapping_mod(pgtbl_t pt, u32_t addr, u32_t flags, u32_t *prevflags)
+{
+	/* Not used for now. TODO: add retypetbl_ref / _deref */
+
```

Comment:
blank line unnecessary. The comment at line 290 should probably be outside?

Reply:
I will fix this